packages/headless/src/controllers/core/generated-answer/headless-searchapi-generated-answer.ts

```diff
@@ -0,0 +1,184 @@
+    }
+    return true;
+  },
+
+  subscribeToSearchRequests: (engine) => {
+    const strictListener = () => {
+      const state = engine.state;
+      const requestId = state.search.requestId;
+      const streamId =
+        state.search.extendedResults.generativeQuestionAnsweringId;
+
+      const genQaEngineId = state.generatedAnswer.id;
+
+      if (
+        subscribeStateManager.engines[genQaEngineId].lastRequestId !== requestId
+      ) {
+        subscribeStateManager.engines[genQaEngineId].lastRequestId = requestId;
+        subscribeStateManager.engines[genQaEngineId].abortController?.abort();
+        engine.dispatch(resetAnswer());
+      }
+
+      const isStreamInProgress =
+        subscribeStateManager.getIsStreamInProgress(genQaEngineId);
+      if (
+        !isStreamInProgress &&
+        streamId &&
+        streamId !== subscribeStateManager.engines[genQaEngineId].lastStreamId
+      ) {
+        subscribeStateManager.engines[genQaEngineId].lastStreamId = streamId;
+        engine.dispatch(
+          streamAnswer({
+            setAbortControllerRef: (ref: AbortController) =>
+              subscribeStateManager.setAbortControllerRef(ref, genQaEngineId),
+          })
+        );
+      }
+    };
+    return engine.subscribe(strictListener);
+  },
+};
+
+export interface SearchAPIGeneratedAnswerAnalyticsClient
+  extends GeneratedAnswerAnalyticsClient {}
+
+/**
+ * Creates a `GeneratedAnswer` controller instance using the search API stream pattern.
+ *
+ * @param engine - The headless engine.
+ * @param props - The configurable `GeneratedAnswer` properties.
+ * @returns A `GeneratedAnswer` controller instance.
+ */
+export function buildSearchAPIGeneratedAnswer(
+  engine: SearchEngine | InsightEngine,
+  analyticsClient: SearchAPIGeneratedAnswerAnalyticsClient,
+  props: GeneratedAnswerProps = {}
+): SearchAPIGeneratedAnswer {
+  if (!loadGeneratedAnswerReducer(engine)) {
+    throw loadReducerError;
+  }
+
+  // const {dispatch} = engine;
+  const controller = buildCoreGeneratedAnswer(engine, analyticsClient, props);
+  const getState = () => engine.state;
+
+  if (!engine.state.generatedAnswer.id) {
+    const genQaEngineId = randomID('genQA-', 12);
+    engine.dispatch(setId({id: genQaEngineId}));
+    subscribeStateManager.engines[genQaEngineId] = {
+      abortController: undefined,
+      lastRequestId: '',
+      lastStreamId: '',
+    };
+  }
+
+  subscribeStateManager.subscribeToSearchRequests(engine);
+
+  const isSearchEngine = (
+    engine: SearchEngine | InsightEngine
+  ): engine is SearchEngine => {
+    return 'executeFirstSearchAfterStandaloneSearchboxRedirect' in engine;
+  };
+
+  return {
+    ...controller,
+
+    get state() {
+      return getState().generatedAnswer;
+    },
+    retry() {
+      if (!isSearchEngine(engine)) {
+        return;
+      }
+      engine.dispatch(
+        executeSearch({
+          legacy: analyticsClient.logRetryGeneratedAnswer(),
+        })
+      );
+    },
```

Comment:
out of curiosity, Is there a particular reason why we are using this pattern of treating the case of the searchEngine in the core controller and returning null for other use cases or it's just a coding preference?

Reply:
this is type narrowing to avoid having to cast the engine type.  I would prefer not doing so while fighting with typescript but the `as unknown as engineType` is even more ugly than this pattern.